sycl/doc/DeviceGlobal.md

```diff
@@ -0,0 +1,658 @@
+                                            // unnamed namespace
+}
+
+namespace sycl::detail {
+namespace {
+
+__sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+  device_global_map::add(&::FuBar,
+    /* mangled name of '::FuBar' */,
+    /* size of underlying 'T' type */,
+    /* bool telling whether variable has 'device_image_scope` property */);
+  device_global_map::add(::__sycl_UNIQUE_STRING,
+    /* mangled name of '::(unnamed)::FuBar' with unique suffix appended */,
+    /* size of underlying 'T' type */,
+    /* bool telling whether variable has 'device_image_scope` property */);
+}
+
+} // namepsace (unnamed)
+} // namespace sycl::detail
+```
+
+The `__sycl_UNIQUE_STRING` variable is defined in the same namespace as the
+second `FuBar` device global, so it can reference the variable through
+unqualified name lookup.  Furthermore, the name of the temporary variable
+(`__sycl_UNIQUE_STRING`) is globally unique, so it is guaranteed not to be
+shadowed by any other name in the translation unit.  This problem with variable
+shadowing is also a problem for the integration footer we use for
+specialization constants.  See the [specialization constant design document][3]
+for more details on this topic.
+
+[3]: <SpecializationConstants.md>
+
+### Changes to the `sycl-post-link` tool
+
+The `sycl-post-link` tool performs its normal algorithm to identify the set of
+kernels and device functions that are bundled together into each module.  Once
+it identifies the functions in each module, it scans those functions looking
+for references to global variables of type `device_global`.  If any device
+global variable decorated with `sycl-device-global-image-life` appears in more
+than one module, the `sycl-post-link` tool issues an error diagnostic:
+
+```
+error: device_global variable <name> with property "device_image_scope"
+       is contained in more than one device image.
+```
+
+Assuming that no error diagnostic is issued, the `sycl-post-link` tool includes
+the IR definition of each `device_global` variable in the modules that
+reference that variable.
+
+The [backend functions described below][4] that allow the host to copy to or
+from a device global require the variable to have "export" linkage in SPIR-V.
+Therefore, the `sycl-post-link` tool needs to make the following IR
+transformations for any `device_global` variable that has internal linkage:
+
+[4]: <#back-end-specific-function-to-copy-to--from-a-device-symbol>
+
+* The linkage is changed to be external.
+* The name of the variable is changed to be the string from the
+  `sycl-unique-id` attribute.
+
+**NOTE**: It seems likely that changing the name of internal linkage variables
+will be bad for debuggability of the code.  The user may attempt to print the
+value of a variable in the debugger, but the debugger won't know the variable
+by that name.  See the "Unresolved issues" section below for more discussion
+on this.
```

Comment:
Note: there are [aliases](https://llvm.org/docs/LangRef.html#aliases) in LLVM IR. I'm not familiar with how they work in details, but probably we could simply introduce an external alias to an internal symbol instead of changing the symbol itself?

Reply:
Actually, this is no longer going to be an issue because we no longer need to rename the variable.  See my comment earlier:

> Actually, I talked with @bashbaug yesterday about resolving the last two unresolved issues at the bottom of this design doc. The solution we agreed to will avoid renaming the user's variable in SPIR-V. We will still need a unique string ID for each device_global variable, but it will no longer be user visible, so there's no impact on debugging.

I'm getting close to resolving the remaining open issues in this design doc, and I plan to make an update soon.  When I do that, I'll also remove the wording about renaming the internal linkage variables.

Reply:
I removed all this wording about renaming variables with internal linkage in da1124837b604f6941e0807ece7cd2292e66f136.  I think we now have a much cleaner way to solve this problem.